pages/people/people.utils.ts

```diff
@@ -0,0 +1,97 @@
+import { DepartmentRecord } from 'types'
+
+// find all ancestors of a department
+export const findParentDepartmentsOfDepartment = (
+	department: DepartmentRecord,
+	allDepartments: DepartmentRecord[]
+): DepartmentRecord[] => {
+	if (!department) {
+		return []
+	}
+
+	const parentDepartments: DepartmentRecord[] = []
+
+	// DFS to find all ancestors of a department
+	const findParentDepartmentsHelper = (
+		cur: DepartmentRecord,
+		path: DepartmentRecord[]
+	) => {
+		if (cur.id === department.id) {
+			parentDepartments.push(...path)
+			return
+		}
+
+		if (cur?.children) {
+			for (const c of cur.children) {
+				findParentDepartmentsHelper(c, [...path, cur])
+			}
+		}
+
+		return
+	}
+
+	allDepartments.forEach((d: DepartmentRecord) =>
+		findParentDepartmentsHelper(d, [])
+	)
+
+	return parentDepartments
+}
+
+// find all descendants of a department
+export const findSubDepartmentsOfDepartment = (
+	department: DepartmentRecord
+): DepartmentRecord[] => {
+	if (!department) {
+		return []
+	}
+
+	const subDepartments: DepartmentRecord[] = []
+
+	const findSubDepartmentsHelper = (cur: DepartmentRecord) => {
+		if (cur.children) {
+			subDepartments.push(...cur.children)
+
+			for (const c of cur.children) {
+				findSubDepartmentsHelper(c)
+			}
+		}
+	}
+
+	findSubDepartmentsHelper(department)
+	return subDepartments
+}
+
+// converts a flat array of departments to a nested array of departments
+export const convertDepartmentsToNested = (
```

Comment:
This algorithm has exponential complexity. How would you refactor so that it had linear or logarithmic complexity instead?

Reply:
You are right, this could be improved (even if it only happens once per refresh of the page). In a real setting, I'd probably ask for advice in the PR on a quicker implementation. 

Thinking about it more, it probably could be done in linear time if for each node, we add it as a child to it's parent. Maybe if we used a map to look up the departments faster, this would be possible.